snuba/consumers/consumer.py

```diff
@@ -384,15 +416,17 @@ def submit(
         assert not self.__closed
 
         for storage_key, payload in message.payload:
-            self.__steps[storage_key].submit(
-                Message(
-                    message.partition,
-                    message.offset,
-                    payload,
-                    message.timestamp,
-                    message.next_offset,
-                )
+            storage_message = Message(
+                message.partition,
+                message.offset,
+                payload,
+                message.timestamp,
+                message.next_offset,
             )
+            self.__messages[storage_key] = self.__messages.get(storage_key, []) + [
+                storage_message
+            ]
```

Comment:
What is this for ?

Reply:
for now I don't have the messages in the `MultistorageCollector` step, so I wasn't sure how to forward them to the next step without collecting them in this one, I could move the produce step further down to like the `InsertBatchWriter` instead

Reply:
@fpacifici @nikhars does this approach make sense to do? In terms of being able to capture the messages that were attempted to be batch inserted

Reply:
I guess it should be ok. I don't know what you intend to put in the dead letter topic. If it's InsertBatch or KafkaMessage. But you can decide that in the DeadLetterStep.

Ideally we should get payload in the dead letter topic in a form so that its easy to apply like an actual INSERT statement. @fpacifici might have opinion on it.

Reply:
I assumed we could have each message that goes into the dead letter topic be a `KafkaPayload`, where maybe the key can be the `storage_key` and the payload could be the rows from the `BytesInsertBatch`. (maybe the payload also has to have the `origin_timestamp` too) But basically I was thinking you could have a consumer strategy that was basically only a `CollectStep`, and you'd pass in something quite similar to the `MultistorageCollector` as the collector.

Reply:
> The flow to me (and correct me if wrong):
collector collects up messages to be batch written -> batch fails -> messages unbatched and produced to dead letter topic -> new consumer collector collects dead letter topic messages to be batch written again -> hopefully batch succeeds

Works for me